src/mlpack/methods/neat/neat.hpp

```diff
@@ -0,0 +1,370 @@
+ *
+ * @tparam TaskType The task NEAT is to be trained on.
+ * @tparam ActivationFunction The activation function used.
+ * @tparam SelectionPolicy The policy used to select genomes.
+ */
+template <class TaskType,
+          class ActivationFunction = ann::HardSigmoidFunction,
+          class SelectionPolicy = RankSelection>
+class NEAT
+{
+ public:
+  /**
+   * Creates a NEAT class instance.
+   *
+   * @param task An instance of the task NEAT is to be trained on.
+   * @param inputNodeCount The number of input nodes of the genomes.
+   * @param outputNodeCount The number of output nodes of the genomes.
+   * @param popSize The size of the population.
+   * @param maxGen The maximum number of generations for which NEAT should run.
+   * @param initialBias The bias with which genomes are initialized.
+   * @param initialWeight The weight of connections with which genomes are 
+   *    initialized.
+   * @param weightMutationProb The probability of a weight being mutated.
+   * @param weightMutationSize The degree of mutation of a weight.
+   * @param biasMutationProb The probability of the bias being mutated.
+   * @param biasMutationSize The degree of mutation of bias.
+   * @param nodeAdditionProb The probability of a new node being added.
+   * @param connAdditionProb The probability of a connection being added.
+   * @param connDeletionProb The probability of a connection being deleted.
+   * @param disableProb The probability of a disabled gene becoming enabled
+   *    during crossover.
+   * @param elitismProp The proportion of a species that is considered elite.
+   * @param finalFitness The desired fitness of the genomes. If it is 0, no
+   *    no limit on the fitness is considered.
+   * @param complexityThreshold The maximum complexity allowed.
+   * @param maxSimplifyGen The maximum number of generations for which
+   *    simplification will occur.
+   * @param isAcyclic Denotes whether or not the genome is meant to be acyclic.
+   */
+  NEAT(TaskType& task,
+       const size_t inputNodeCount,
+       const size_t outputNodeCount,
+       const size_t popSize,
+       const size_t maxGen,
+       const size_t numSpecies,
+       const double initialBias = 1,
+       const double initialWeight = 0,
+       const double weightMutationProb = 0.8,
+       const double weightMutationSize = 0.5,
+       const double biasMutationProb = 0.7,
+       const double biasMutationSize = 0.5,
+       const double nodeAdditionProb = 0.2,
+       const double connAdditionProb = 0.5,
+       const double connDeletionProb = 0.5,
+       const double disableProb = 0.2,
+       const double elitismProp = 0.1,
+       const double finalFitness = 0,
+       const size_t complexityThreshold = 0,
+       const size_t maxSimplifyGen = 10,
+       const bool isAcyclic = false);
+
+  /**
+   * Trains the model on the task and returns the best Genome.
+   */
+  Genome<ActivationFunction> Train();
```

Comment:
To be consistent with the rest of the codebase I think it would make sense to pass a reference of the genome.